crates/runtime/src/memory.rs

```diff
@@ -9,9 +9,38 @@ use std::cell::RefCell;
 use std::convert::TryFrom;
 use wasmtime_environ::{MemoryPlan, MemoryStyle, WASM_MAX_PAGES, WASM_PAGE_SIZE};
 
+/// A memory allocator
+pub trait Allocator: Send + Sync {
+    /// Create new LinearMemory
+    fn new_memory(&self, plan: &MemoryPlan) -> Result<Box<dyn LinearMemory>, String>;
+}
+
+/// A default memory allocator used by Wasmtime
+pub struct DefaultAllocator;
+
+impl Allocator for DefaultAllocator {
+    /// Create new MmapMemory
+    fn new_memory(&self, plan: &MemoryPlan) -> Result<Box<dyn LinearMemory>, String> {
+        Ok(Box::new(MmapMemory::new(plan)?) as Box<dyn LinearMemory>)
+    }
+}
+
+/// A linear memory
+pub trait LinearMemory {
```

Comment:
In terms of safety *something* here needs to be `unsafe` due to the trickiness to get this all right (and segfaults which happen if there's a bug). I think all traits introduced here (both `Allocator` and `LinearMemory`) will need to be `unsafe trait` declarations since they have a number of safety guarantees to uphold in their implementations.